### ISSUE TITLE
texture builtins: For most texture builtins, Vulkan only allows compile-time-constant offset, and with with bounded range

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#1235]
The WGSL texture functions have an offset parameter of signed integral type (with same number of components as the coordinate).  So far these are not otherwise constrained.

But Vulkan requires these to be compile-time constants.

Also, implementations have defined limits minTexelOffset (no higher than -8)  and maxTexelOffset (no lower than 7)
See the [Required Limits table](https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#limits-minmax).
Metal/MSL also limit the range from -8 through 7+.

Background: The offset operand in texture functions for Vulkan/SPIR-V has a few limitations that we need to account for:

- There are 3 relevant image operands in SPIR-V
     - `Offset`: allows non-constant integer coordinate offset.
              -   Vulkan limits this to only the OpImageGather and OpImageDrefGather instructions.  See [Vulkan spir-v environment](https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#spirvenv-module-validation-standalone):  "Image operand Offset must only be used with OpImage*Gather instructions."
     - `ConstOffsets`: allows array of compile-time-constant offsets for Gather and DrefGather.  
     - `ConstOffset` :   usable by many image instructions.   This only allows a compile-time-constant offset.

@ben-clayton

**Comment by dneto0** [comment]
We could require the operand to be a const_expr.  In this sense the texture builtins are not like ordinary function calls.

**Comment by dneto0** [comment]
@ben-clayton pointed out that the -8 to +7 limit is already captured.

**Comment by ben-clayton** [comment]
The limits on the offset [are already documented](https://gpuweb.github.io/gpuweb/wgsl.html#texturesample), but you're right that offsets need to be compile time constant and this is not documented. 

I like your suggestion of needing the offset to be `const_expr`, but yes, this is a bit special. Unless someone can think up a better suggestion, I'll draft a PR to update the spec tomorrow.

**Comment by kvark** [comment]
> See Vulkan spir-v environment: 

That's the annoying thing Apple pointed to on one of the past F2F conferences - it's hard to navigate the spec when it's split over multiple documents like this one.
Would it be totally nuts to ask for some sort of a combined specification of SPIR-V? I.e one that takes SPIR-V proper and adds the Vulkan environment constraints **inline**, so that we (who work on WGSL) can see it and also reason about it better?

> const_expr

Do we even have const expr? I recall asking about this semantics in #905 and the resolution was that constants aren't computed, they have to be literals.

**Comment by dj2** [comment]
We have a grammar declaration named `const_expr`:

```
const_expr
  : type_decl PAREN_LEFT (const_expr COMMA)* const_expr PAREN_RIGHT
  | const_literal

const_literal
  : INT_LITERAL
  | UINT_LITERAL
  | FLOAT_LITERAL
  | TRUE
  | FALSE
```

**Comment by kvark** [comment]
In that sense, yes, looks like we have no other choice but to change the grammar to expect the `const_expr` for the offsets.

**Comment by ben-clayton** [comment]
I've drafted a spec update at: https://github.com/gpuweb/gpuweb/pull/1238

**Comment by jdashg** [comment]
Fixed by #1238.

**Comment by dneto0** [comment]
FYI.     Vulkan allows GLSL textureFetchOffset on sampled images.  That translates to SPIR-V OpImageFetch with ConstOffset parameter.  

OpImageFetch maps to WGSL textureLoad.
But textureLoad does not have overloads with a `offset` parameter.  That's because Metal doesn't provide that.

E.g. for 2D textures:

> These member functions perform sampler-less reads from a 2D texture:
> Tv read(uint2 coord, uint lod = 0) const
> Tv read(ushort2 coord,  ushort lod = 0) const // All OS: Metal 1.2 and later.

That does not include a const 'offset' parameter.  The shader has to fold it into the coordinate parameter.
